module-3/lab-supervised-learning-feature-extraction/your-code/main.ipynb

```diff
@@ -483,7 +483,7 @@
    "outputs": [],
```

Comment:
- In this one "We see that Google Play is using a shorthand for millions.", a more simple way:
```Python 
def convert_string_to_numeric(s):
    if s[-1] == 'M':
        return float(s[:-1]) * 1000000
    else:
        return float(s)
    return np.NaN
test_string = '4.0M'
convert_string_to_numeric(test_string) == 4000000
```
- Remember to apply format method in case you want to avoid scientific notation.
- In this ones "While we can convert most of the Size values to numeric in the same way we converted the Reviews values, there is one value that is impossible to convert." other way of doing it:
```Python
odd_value = "Varies with device"
google_play[google_play.Size == odd_value].shape[0] / google_play.shape[0]
```
- In this one "What is the proportion of the missing values in Rating to the total number of records?" other way using shape method.
```Python
#your code
a = (nulls['Rating']/ len(google_play))*100
a
#Using shape and sum methods
google_play.isna().sum()['Rating'] / google_play.shape[0]
```
- You don't need to assign to a variable in this one : 
```Python 
#your code
a =pd.to_datetime(google_missing_removed['Last Updated'])
google_missing_removed['Last Updated']= a
#More pythonic way: 
google_missing_removed['Last Updated'] = pd.to_datetime(google_missing_removed['Last Updated'])
```
- You can do this in a single line: 
```Python 
#Your code 
price_n = []
for i in google_missing_removed['Price']:
    i = i.replace('$', '')
    price_n.append(i)
google_missing_removed['Price'] = price_n
#Single line example 
google_missing_removed['Price Numerical'] = google_missing_removed['Price'].str.replace('$', '')
```
- In this one "Now, use the value_counts() function (documentation) to get a sense on how many apps are in this dataset and their review count" something as simple as the following also works:
```Python
review_missing_removed.App.value_counts()
```
- Good one 👍 : 
```Python 
#Your code 
def positive_function(x):
        return sum(e == 'Positive' for e in x)
#Other way of doing the same stuff 
def positive_function(x):
    return sum(np.where(x == "Positive", 1, 0))
```
- "Bonus": you can do the following way to rename columns: 
```Python 
#Your code 
google_agg = review_missing_removed.groupby('App').agg({'Sentiment': [positive_function, len]})
google_agg.columns = ['Positive', 'Total']
#An alternative 
google_agg = review_missing_removed.groupby('App').Sentiment.agg([positive_function, 'count']).rename({'positive_function': 'Positive', 'count': 'Total'}, axis='columns')
```
- The key here "Add a derived column to google_agg that is the ratio of the Positive and the Total columns. Call this column Positive Ratio.Make sure to account for the case where the denominator is zero using the np.where() function." is to consider a zero denominator. Your code doesn't consider that case: 
```Python 
#Your code
google_agg['Positive_Ratio'] = google_agg['Positive'] / google_agg['Total']
#One way that considers a zero denominator 
google_agg['Positive Ratio'] = np.where(google_agg['Total'] !=0, google_agg['Positive'] / google_agg['Total'], 0)
```
- Good job trying to train a SVM model but the positive ratio can't be considered a classification problem. What are the classes? None. In case it has sense a problem on positive ratio as the target, it could be considered more a regression problem than a classification problem. 
- Some addtional things:
1. When training a classifier there's one important thing you should look at because it can cause unexpected and unwanted results, are the classes balanced or unbalanced in your dataset? (Imagine you have 2000 1 and 550 0 records for your target) Ask Marc to explain the consequences of unbalanced classes in models performance. Take a look on the 2 links below but there are many more, feel free to explore on this question.
https://machinelearningmastery.com/tactics-to-combat-imbalanced-classes-in-your-machine-learning-dataset/
https://towardsdatascience.com/dealing-with-imbalanced-classes-in-machine-learning-d43d6fa19d2
2. In case we have features in different scales we often standardize the numeric variables. A common method to make all features comparable among them is subtracting the feature mean and dividing the feature standard deviation to each record in the data set. This can be do using the standard scaler of sklearn: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
3. When training a classifier there's one important thing you should look at because it can cause unexpected and unwanted results.Are the  classes balanced or unbalanced in your dataset? (Imagine you have 2000 1 and 550 0 records for your target) Ask Marc to explain the consequences of unbalanced classes in the models performance. Take a look on the 2 links below but there are many more, feel free to explore on this question.
https://machinelearningmastery.com/tactics-to-combat-imbalanced-classes-in-your-machine-learning-dataset/
https://towardsdatascience.com/dealing-with-imbalanced-classes-in-machine-learning-d43d6fa19d2
4. Other important thing before training any model is to ensure that outliers have been removed or preprocessed accordingly.